# lgbm-tfidf 도전

In [9]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline
import lightgbm as lgb
from matplotlib import pyplot as plt
from matplotlib import rcParams
import numpy as np
from pathlib import Path
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import StratifiedKFold
import seaborn as sns
import warnings
from nltk.tokenize import word_tokenize
from nltk.tokenize import WordPunctTokenizer
from nltk.tokenize.regexp import RegexpTokenizer
from nltk.stem.snowball import SnowballStemmer 
import nltk
from nltk.corpus import stopwords
from sklearn.metrics import accuracy_score, log_loss
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer
import os

In [10]:
warnings.filterwarnings(action='ignore')
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.preprocessing.text import Tokenizer
import re

In [16]:
path = './data'
trn = pd.read_csv(os.path.join(path,'train.csv'),encoding = 'utf-8')
tst = pd.read_csv(os.path.join(path,'test_x.csv'),encoding = 'utf-8')
sample_submission = pd.read_csv(os.path.join(path,'sample_submission.csv'),encoding = 'utf-8')

#trn = pd.read_csv('C:\\Users\\USER\\Desktop\\open\\train.csv', encoding = 'utf-8')
#tst = pd.read_csv('C:\\Users\\USER\\Desktop\\open\\test_x.csv', encoding = 'utf-8')
#sample_submission = pd.read_csv('C:\\Users\\USER\\Desktop\\open\\sample_submission.csv', encoding = 'utf-8')

[문자열 가공]  
1. Bag-of-Words(simple):countVectorizer, TfidVectorizer  
2. Hashing Trick: hashingVectorizer  
3. Embeddings(with 신경망): Word2Vec, GloVe

## Bag-of-Words_Count 피처 생성1

In [27]:
vec = CountVectorizer(tokenizer=word_tokenize, stop_words=stopwords.words('english'), ngram_range=(1, 3), min_df=100)
X_cnt = vec.fit_transform(trn['text'])
X_tst = vec.fit_transform(tst['text'])
print(X_cnt.shape, X_tst.shape)

(54879, 2892) (19617, 1990)


In [28]:
X_cnt[0, :50].todense()

matrix([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0]], dtype=int64)

In [30]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

## Lgbm 모델 학습_count

In [33]:
y=trn.author.values
X_cnt = X_cnt.astype('float32') 
y = y.astype('float32')
X_tst = X_tst.astype('float32')   
p = np.zeros((X_cnt.shape[0], 5))
p_tst = np.zeros((X_tst.shape[0], 5))
#(cv.split(),1)
# -  StratifiedKFold 경우, trn,y 독립과 종속(동일 분포를 위해) 인자
for i, (i_trn, i_val) in enumerate(cv.split(X_cnt, y), 1):
    print(f'training model for CV #{i}')
    clf = lgb.LGBMClassifier(objective='multiclass',
                             n_estimators=3000,
                             num_leaves=256,
                             learning_rate=0.009,
                             subsample=.8,
                             subsample_freq=1,
                             num_iterations=2000,
                             colsample_bytree=.6,
                             random_state=42,
                             n_jobs=-1)
    
    clf.fit(X_cnt[i_trn], y[i_trn],eval_set=[(X_cnt[i_val], y[i_val])], eval_metric='multiclass',early_stopping_rounds=15)
    p[i_val, :] = clf.predict_proba(X_cnt[i_val])
    #확률로 예측값 출력
    p_tst += clf.predict_proba(X_tst) / 5
    #평균 효과를 위해 먼저 폴드 개수만큼 나누어 초기화

training model for CV #1
[1]	valid_0's multi_logloss: 1.55939
Training until validation scores don't improve for 15 rounds
[2]	valid_0's multi_logloss: 1.54824
[3]	valid_0's multi_logloss: 1.53794
[4]	valid_0's multi_logloss: 1.52787
[5]	valid_0's multi_logloss: 1.51826
[6]	valid_0's multi_logloss: 1.50833
[7]	valid_0's multi_logloss: 1.49876
[8]	valid_0's multi_logloss: 1.48943
[9]	valid_0's multi_logloss: 1.48034
[10]	valid_0's multi_logloss: 1.47112
[11]	valid_0's multi_logloss: 1.46225
[12]	valid_0's multi_logloss: 1.45356
[13]	valid_0's multi_logloss: 1.44575
[14]	valid_0's multi_logloss: 1.43761
[15]	valid_0's multi_logloss: 1.42952
[16]	valid_0's multi_logloss: 1.4214
[17]	valid_0's multi_logloss: 1.41366
[18]	valid_0's multi_logloss: 1.40599
[19]	valid_0's multi_logloss: 1.39846
[20]	valid_0's multi_logloss: 1.39101
[21]	valid_0's multi_logloss: 1.38326
[22]	valid_0's multi_logloss: 1.37611
[23]	valid_0's multi_logloss: 1.36894
[24]	valid_0's multi_logloss: 1.36176
[25]	valid_0

[210]	valid_0's multi_logloss: 0.849276
[211]	valid_0's multi_logloss: 0.848168
[212]	valid_0's multi_logloss: 0.847028
[213]	valid_0's multi_logloss: 0.845886
[214]	valid_0's multi_logloss: 0.844742
[215]	valid_0's multi_logloss: 0.84366
[216]	valid_0's multi_logloss: 0.842571
[217]	valid_0's multi_logloss: 0.841456
[218]	valid_0's multi_logloss: 0.840341
[219]	valid_0's multi_logloss: 0.839218
[220]	valid_0's multi_logloss: 0.838215
[221]	valid_0's multi_logloss: 0.837141
[222]	valid_0's multi_logloss: 0.836113
[223]	valid_0's multi_logloss: 0.83497
[224]	valid_0's multi_logloss: 0.83396
[225]	valid_0's multi_logloss: 0.833002
[226]	valid_0's multi_logloss: 0.831999
[227]	valid_0's multi_logloss: 0.830953
[228]	valid_0's multi_logloss: 0.829918
[229]	valid_0's multi_logloss: 0.828955
[230]	valid_0's multi_logloss: 0.827927
[231]	valid_0's multi_logloss: 0.826883
[232]	valid_0's multi_logloss: 0.825871
[233]	valid_0's multi_logloss: 0.824903
[234]	valid_0's multi_logloss: 0.823984
[23

[416]	valid_0's multi_logloss: 0.705618
[417]	valid_0's multi_logloss: 0.705175
[418]	valid_0's multi_logloss: 0.704705
[419]	valid_0's multi_logloss: 0.704342
[420]	valid_0's multi_logloss: 0.703914
[421]	valid_0's multi_logloss: 0.703507
[422]	valid_0's multi_logloss: 0.703137
[423]	valid_0's multi_logloss: 0.702743
[424]	valid_0's multi_logloss: 0.702361
[425]	valid_0's multi_logloss: 0.701961
[426]	valid_0's multi_logloss: 0.701535
[427]	valid_0's multi_logloss: 0.701114
[428]	valid_0's multi_logloss: 0.700793
[429]	valid_0's multi_logloss: 0.700445
[430]	valid_0's multi_logloss: 0.700071
[431]	valid_0's multi_logloss: 0.699699
[432]	valid_0's multi_logloss: 0.699283
[433]	valid_0's multi_logloss: 0.698904
[434]	valid_0's multi_logloss: 0.698501
[435]	valid_0's multi_logloss: 0.698096
[436]	valid_0's multi_logloss: 0.69768
[437]	valid_0's multi_logloss: 0.697247
[438]	valid_0's multi_logloss: 0.696835
[439]	valid_0's multi_logloss: 0.69645
[440]	valid_0's multi_logloss: 0.696099
[4

[622]	valid_0's multi_logloss: 0.649197
[623]	valid_0's multi_logloss: 0.649014
[624]	valid_0's multi_logloss: 0.648863
[625]	valid_0's multi_logloss: 0.648722
[626]	valid_0's multi_logloss: 0.648549
[627]	valid_0's multi_logloss: 0.648396
[628]	valid_0's multi_logloss: 0.648204
[629]	valid_0's multi_logloss: 0.648054
[630]	valid_0's multi_logloss: 0.647924
[631]	valid_0's multi_logloss: 0.647769
[632]	valid_0's multi_logloss: 0.647621
[633]	valid_0's multi_logloss: 0.64749
[634]	valid_0's multi_logloss: 0.647323
[635]	valid_0's multi_logloss: 0.647164
[636]	valid_0's multi_logloss: 0.646982
[637]	valid_0's multi_logloss: 0.646814
[638]	valid_0's multi_logloss: 0.646664
[639]	valid_0's multi_logloss: 0.64651
[640]	valid_0's multi_logloss: 0.646346
[641]	valid_0's multi_logloss: 0.646174
[642]	valid_0's multi_logloss: 0.645972
[643]	valid_0's multi_logloss: 0.645829
[644]	valid_0's multi_logloss: 0.645632
[645]	valid_0's multi_logloss: 0.645477
[646]	valid_0's multi_logloss: 0.645329
[6

[828]	valid_0's multi_logloss: 0.62565
[829]	valid_0's multi_logloss: 0.625582
[830]	valid_0's multi_logloss: 0.625522
[831]	valid_0's multi_logloss: 0.625455
[832]	valid_0's multi_logloss: 0.625428
[833]	valid_0's multi_logloss: 0.625355
[834]	valid_0's multi_logloss: 0.625279
[835]	valid_0's multi_logloss: 0.62525
[836]	valid_0's multi_logloss: 0.625198
[837]	valid_0's multi_logloss: 0.625128
[838]	valid_0's multi_logloss: 0.625071
[839]	valid_0's multi_logloss: 0.624999
[840]	valid_0's multi_logloss: 0.624933
[841]	valid_0's multi_logloss: 0.624893
[842]	valid_0's multi_logloss: 0.624839
[843]	valid_0's multi_logloss: 0.62478
[844]	valid_0's multi_logloss: 0.624758
[845]	valid_0's multi_logloss: 0.624701
[846]	valid_0's multi_logloss: 0.624661
[847]	valid_0's multi_logloss: 0.624612
[848]	valid_0's multi_logloss: 0.624537
[849]	valid_0's multi_logloss: 0.624464
[850]	valid_0's multi_logloss: 0.624392
[851]	valid_0's multi_logloss: 0.624353
[852]	valid_0's multi_logloss: 0.624275
[85

[1033]	valid_0's multi_logloss: 0.616826
[1034]	valid_0's multi_logloss: 0.616798
[1035]	valid_0's multi_logloss: 0.616789
[1036]	valid_0's multi_logloss: 0.61676
[1037]	valid_0's multi_logloss: 0.616743
[1038]	valid_0's multi_logloss: 0.616721
[1039]	valid_0's multi_logloss: 0.616704
[1040]	valid_0's multi_logloss: 0.616668
[1041]	valid_0's multi_logloss: 0.616645
[1042]	valid_0's multi_logloss: 0.616638
[1043]	valid_0's multi_logloss: 0.616612
[1044]	valid_0's multi_logloss: 0.616592
[1045]	valid_0's multi_logloss: 0.616579
[1046]	valid_0's multi_logloss: 0.616592
[1047]	valid_0's multi_logloss: 0.61656
[1048]	valid_0's multi_logloss: 0.616535
[1049]	valid_0's multi_logloss: 0.616512
[1050]	valid_0's multi_logloss: 0.616506
[1051]	valid_0's multi_logloss: 0.616478
[1052]	valid_0's multi_logloss: 0.616475
[1053]	valid_0's multi_logloss: 0.616458
[1054]	valid_0's multi_logloss: 0.61643
[1055]	valid_0's multi_logloss: 0.616406
[1056]	valid_0's multi_logloss: 0.616375
[1057]	valid_0's mu

ValueError: Number of features of the model must match the input. Model n_features_ is 2892 and input n_features is 1990 

In [34]:
print(f'Accuracy (CV): {accuracy_score(y, np.argmax(p, axis=1)) * 100:8.4f}%')
print(f'Log Loss (CV): {log_loss(pd.get_dummies(y), p):8.4f}')

Accuracy (CV):  34.5542%
Log Loss (CV):   1.4107


## Bag-of-Words_Tfidf 피처 생성2

In [20]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/silbu73/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/silbu73/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [21]:
vec = TfidfVectorizer(tokenizer=word_tokenize, stop_words=stopwords.words('english'), ngram_range=(1, 3), min_df=50)
X_tfi = vec.fit_transform(trn['text'])
X_tst = vec.transform(tst['text'])
print(X_tfi.shape, X_tst.shape)

(54879, 5897) (19617, 5897)


In [22]:
X_tfi[0, :50].todense()

matrix([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0.]])

In [23]:
y=trn.author.values
X_tfi = X_tfi.astype('float32') 
y = y.astype('float32')
X_tst = X_tst.astype('float32')

## 하이퍼 파라미터 튜닝

In [24]:
from hyperopt import STATUS_OK, Trials, hp, space_eval, tpe, fmin
from sklearn.model_selection import StratifiedKFold, train_test_split

In [26]:
X_trn, X_val, y_trn, y_val = train_test_split(X_tfi, y, test_size=.2, random_state=42)

In [28]:
params = {
    "objective": "multiclass",
    "n_estimators": 1000,
    "subsample_freq": 1,
    "random_state": 42,
    "n_jobs": -1,
}

space = {
    "learning_rate": hp.loguniform("learning_rate", np.log(0.009), np.log(0.3)),
    "num_leaves": hp.choice("num_leaves", [15, 31, 63, 127, 255]),
    "colsample_bytree": hp.quniform("colsample_bytree", .4, .9, 0.1),
    "subsample": hp.quniform("subsample", .4, .9, 0.1),
}

In [29]:
def objective(hyperparams):
    model = lgb.LGBMClassifier(**params, **hyperparams)
    model.fit(X=X_trn, y=y_trn,
              eval_set=[(X_val, y_val)],
              eval_metric="multi_logloss",
              early_stopping_rounds=10,
              verbose=False)
    score = model.best_score_["valid_0"]["multi_logloss"]

    return {'loss': score, 'status': STATUS_OK, 'model': model}

trials = Trials()
best = fmin(fn=objective, space=space, trials=trials,
            algo=tpe.suggest, max_evals=10, verbose=1)

hyperparams = space_eval(space, best)
n_best = trials.best_trial['result']['model'].best_iteration_
params.update(hyperparams)
print(params)

100%|██████████| 10/10 [24:27<00:00, 146.75s/trial, best loss: 0.6236845738732151]
{'objective': 'multiclass', 'n_estimators': 1000, 'subsample_freq': 1, 'random_state': 42, 'n_jobs': -1, 'colsample_bytree': 0.8, 'learning_rate': 0.04147603536728929, 'num_leaves': 127, 'subsample': 0.9}


## Stratified K-Fold Cross Validation

In [30]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

## Lgbm 모델 학습_tfidf

In [31]:
p = np.zeros((X_tfi.shape[0], 5))
p_tst = np.zeros((X_tst.shape[0], 5))
#(cv.split(),1)
# -  StratifiedKFold 경우, trn,y 독립과 종속(동일 분포를 위해) 인자
for i, (i_trn, i_val) in enumerate(cv.split(X_tfi, y), 1):
    print(f'training model for CV #{i}')
    clf = lgb.LGBMClassifier(objective='multiclass',
                             n_estimators=1000,
                             num_leaves=127,
                             learning_rate=0.04147603536728929,
                             subsample=.9,
                             subsample_freq=1,
                             num_iterations=2000,
                             colsample_bytree=.8,
                             random_state=42,
                             n_jobs=-1)
    
    clf.fit(X_tfi[i_trn], y[i_trn],eval_set=[(X_tfi[i_val], y[i_val])], eval_metric='multiclass')
    #early_stopping_rounds=15
    p[i_val, :] = clf.predict_proba(X_tfi[i_val])
    #확률로 예측값 출력
    p_tst += clf.predict_proba(X_tst) / 5
    #평균 효과를 위해 먼저 폴드 개수만큼 나누어 초기화

training model for CV #1
[1]	valid_0's multi_logloss: 1.51853
[2]	valid_0's multi_logloss: 1.47259
[3]	valid_0's multi_logloss: 1.4322
[4]	valid_0's multi_logloss: 1.39536
[5]	valid_0's multi_logloss: 1.36216
[6]	valid_0's multi_logloss: 1.33023
[7]	valid_0's multi_logloss: 1.30199
[8]	valid_0's multi_logloss: 1.27433
[9]	valid_0's multi_logloss: 1.24921
[10]	valid_0's multi_logloss: 1.22659
[11]	valid_0's multi_logloss: 1.2053
[12]	valid_0's multi_logloss: 1.18497
[13]	valid_0's multi_logloss: 1.16609
[14]	valid_0's multi_logloss: 1.14904
[15]	valid_0's multi_logloss: 1.13281
[16]	valid_0's multi_logloss: 1.11681
[17]	valid_0's multi_logloss: 1.10185
[18]	valid_0's multi_logloss: 1.08852
[19]	valid_0's multi_logloss: 1.07499
[20]	valid_0's multi_logloss: 1.06167
[21]	valid_0's multi_logloss: 1.04933
[22]	valid_0's multi_logloss: 1.03755
[23]	valid_0's multi_logloss: 1.02631
[24]	valid_0's multi_logloss: 1.01568
[25]	valid_0's multi_logloss: 1.00536
[26]	valid_0's multi_logloss: 0.9959

[209]	valid_0's multi_logloss: 0.623669
[210]	valid_0's multi_logloss: 0.62338
[211]	valid_0's multi_logloss: 0.623047
[212]	valid_0's multi_logloss: 0.622788
[213]	valid_0's multi_logloss: 0.622633
[214]	valid_0's multi_logloss: 0.622399
[215]	valid_0's multi_logloss: 0.622182
[216]	valid_0's multi_logloss: 0.621928
[217]	valid_0's multi_logloss: 0.621654
[218]	valid_0's multi_logloss: 0.621347
[219]	valid_0's multi_logloss: 0.621139
[220]	valid_0's multi_logloss: 0.620782
[221]	valid_0's multi_logloss: 0.620531
[222]	valid_0's multi_logloss: 0.620305
[223]	valid_0's multi_logloss: 0.620046
[224]	valid_0's multi_logloss: 0.619708
[225]	valid_0's multi_logloss: 0.619575
[226]	valid_0's multi_logloss: 0.619332
[227]	valid_0's multi_logloss: 0.619064
[228]	valid_0's multi_logloss: 0.618829
[229]	valid_0's multi_logloss: 0.618759
[230]	valid_0's multi_logloss: 0.618612
[231]	valid_0's multi_logloss: 0.618383
[232]	valid_0's multi_logloss: 0.618281
[233]	valid_0's multi_logloss: 0.618033
[

[415]	valid_0's multi_logloss: 0.614942
[416]	valid_0's multi_logloss: 0.615044
[417]	valid_0's multi_logloss: 0.615089
[418]	valid_0's multi_logloss: 0.615122
[419]	valid_0's multi_logloss: 0.615129
[420]	valid_0's multi_logloss: 0.615308
[421]	valid_0's multi_logloss: 0.615362
[422]	valid_0's multi_logloss: 0.615496
[423]	valid_0's multi_logloss: 0.615574
[424]	valid_0's multi_logloss: 0.615587
[425]	valid_0's multi_logloss: 0.615672
[426]	valid_0's multi_logloss: 0.615628
[427]	valid_0's multi_logloss: 0.615748
[428]	valid_0's multi_logloss: 0.615922
[429]	valid_0's multi_logloss: 0.616028
[430]	valid_0's multi_logloss: 0.616151
[431]	valid_0's multi_logloss: 0.616216
[432]	valid_0's multi_logloss: 0.616428
[433]	valid_0's multi_logloss: 0.616569
[434]	valid_0's multi_logloss: 0.616652
[435]	valid_0's multi_logloss: 0.616783
[436]	valid_0's multi_logloss: 0.616856
[437]	valid_0's multi_logloss: 0.616929
[438]	valid_0's multi_logloss: 0.617158
[439]	valid_0's multi_logloss: 0.617274


[621]	valid_0's multi_logloss: 0.640706
[622]	valid_0's multi_logloss: 0.640864
[623]	valid_0's multi_logloss: 0.641038
[624]	valid_0's multi_logloss: 0.641355
[625]	valid_0's multi_logloss: 0.641437
[626]	valid_0's multi_logloss: 0.641713
[627]	valid_0's multi_logloss: 0.641915
[628]	valid_0's multi_logloss: 0.642125
[629]	valid_0's multi_logloss: 0.642359
[630]	valid_0's multi_logloss: 0.64251
[631]	valid_0's multi_logloss: 0.642725
[632]	valid_0's multi_logloss: 0.64275
[633]	valid_0's multi_logloss: 0.642955
[634]	valid_0's multi_logloss: 0.643086
[635]	valid_0's multi_logloss: 0.643278
[636]	valid_0's multi_logloss: 0.643481
[637]	valid_0's multi_logloss: 0.643718
[638]	valid_0's multi_logloss: 0.643889
[639]	valid_0's multi_logloss: 0.643995
[640]	valid_0's multi_logloss: 0.644154
[641]	valid_0's multi_logloss: 0.644439
[642]	valid_0's multi_logloss: 0.644625
[643]	valid_0's multi_logloss: 0.64477
[644]	valid_0's multi_logloss: 0.64506
[645]	valid_0's multi_logloss: 0.64525
[646]

[827]	valid_0's multi_logloss: 0.678958
[828]	valid_0's multi_logloss: 0.679148
[829]	valid_0's multi_logloss: 0.679282
[830]	valid_0's multi_logloss: 0.679431
[831]	valid_0's multi_logloss: 0.679667
[832]	valid_0's multi_logloss: 0.680007
[833]	valid_0's multi_logloss: 0.680207
[834]	valid_0's multi_logloss: 0.680383
[835]	valid_0's multi_logloss: 0.680667
[836]	valid_0's multi_logloss: 0.680882
[837]	valid_0's multi_logloss: 0.681196
[838]	valid_0's multi_logloss: 0.681421
[839]	valid_0's multi_logloss: 0.681634
[840]	valid_0's multi_logloss: 0.681761
[841]	valid_0's multi_logloss: 0.682007
[842]	valid_0's multi_logloss: 0.682208
[843]	valid_0's multi_logloss: 0.68244
[844]	valid_0's multi_logloss: 0.682632
[845]	valid_0's multi_logloss: 0.682837
[846]	valid_0's multi_logloss: 0.682997
[847]	valid_0's multi_logloss: 0.683297
[848]	valid_0's multi_logloss: 0.683527
[849]	valid_0's multi_logloss: 0.683723
[850]	valid_0's multi_logloss: 0.683942
[851]	valid_0's multi_logloss: 0.684128
[

[1032]	valid_0's multi_logloss: 0.72194
[1033]	valid_0's multi_logloss: 0.722186
[1034]	valid_0's multi_logloss: 0.722355
[1035]	valid_0's multi_logloss: 0.722623
[1036]	valid_0's multi_logloss: 0.722816
[1037]	valid_0's multi_logloss: 0.723045
[1038]	valid_0's multi_logloss: 0.723256
[1039]	valid_0's multi_logloss: 0.723492
[1040]	valid_0's multi_logloss: 0.72375
[1041]	valid_0's multi_logloss: 0.723922
[1042]	valid_0's multi_logloss: 0.724172
[1043]	valid_0's multi_logloss: 0.724367
[1044]	valid_0's multi_logloss: 0.724535
[1045]	valid_0's multi_logloss: 0.724728
[1046]	valid_0's multi_logloss: 0.724876
[1047]	valid_0's multi_logloss: 0.725053
[1048]	valid_0's multi_logloss: 0.725298
[1049]	valid_0's multi_logloss: 0.725526
[1050]	valid_0's multi_logloss: 0.725733
[1051]	valid_0's multi_logloss: 0.725949
[1052]	valid_0's multi_logloss: 0.726106
[1053]	valid_0's multi_logloss: 0.726297
[1054]	valid_0's multi_logloss: 0.726492
[1055]	valid_0's multi_logloss: 0.72679
[1056]	valid_0's mu

[1233]	valid_0's multi_logloss: 0.767174
[1234]	valid_0's multi_logloss: 0.767466
[1235]	valid_0's multi_logloss: 0.767662
[1236]	valid_0's multi_logloss: 0.767795
[1237]	valid_0's multi_logloss: 0.767996
[1238]	valid_0's multi_logloss: 0.768212
[1239]	valid_0's multi_logloss: 0.768425
[1240]	valid_0's multi_logloss: 0.768667
[1241]	valid_0's multi_logloss: 0.768969
[1242]	valid_0's multi_logloss: 0.769133
[1243]	valid_0's multi_logloss: 0.76933
[1244]	valid_0's multi_logloss: 0.769624
[1245]	valid_0's multi_logloss: 0.76978
[1246]	valid_0's multi_logloss: 0.769989
[1247]	valid_0's multi_logloss: 0.770218
[1248]	valid_0's multi_logloss: 0.770515
[1249]	valid_0's multi_logloss: 0.770777
[1250]	valid_0's multi_logloss: 0.771067
[1251]	valid_0's multi_logloss: 0.771326
[1252]	valid_0's multi_logloss: 0.771624
[1253]	valid_0's multi_logloss: 0.771902
[1254]	valid_0's multi_logloss: 0.772111
[1255]	valid_0's multi_logloss: 0.772507
[1256]	valid_0's multi_logloss: 0.772775
[1257]	valid_0's m

[1434]	valid_0's multi_logloss: 0.815917
[1435]	valid_0's multi_logloss: 0.816176
[1436]	valid_0's multi_logloss: 0.816444
[1437]	valid_0's multi_logloss: 0.816734
[1438]	valid_0's multi_logloss: 0.81693
[1439]	valid_0's multi_logloss: 0.817249
[1440]	valid_0's multi_logloss: 0.817502
[1441]	valid_0's multi_logloss: 0.817725
[1442]	valid_0's multi_logloss: 0.817932
[1443]	valid_0's multi_logloss: 0.818102
[1444]	valid_0's multi_logloss: 0.818321
[1445]	valid_0's multi_logloss: 0.818631
[1446]	valid_0's multi_logloss: 0.818985
[1447]	valid_0's multi_logloss: 0.819181
[1448]	valid_0's multi_logloss: 0.819412
[1449]	valid_0's multi_logloss: 0.819577
[1450]	valid_0's multi_logloss: 0.819806
[1451]	valid_0's multi_logloss: 0.819959
[1452]	valid_0's multi_logloss: 0.820196
[1453]	valid_0's multi_logloss: 0.820456
[1454]	valid_0's multi_logloss: 0.820762
[1455]	valid_0's multi_logloss: 0.821082
[1456]	valid_0's multi_logloss: 0.821289
[1457]	valid_0's multi_logloss: 0.821512
[1458]	valid_0's 

[1635]	valid_0's multi_logloss: 0.86598
[1636]	valid_0's multi_logloss: 0.866223
[1637]	valid_0's multi_logloss: 0.866479
[1638]	valid_0's multi_logloss: 0.86675
[1639]	valid_0's multi_logloss: 0.866988
[1640]	valid_0's multi_logloss: 0.867331
[1641]	valid_0's multi_logloss: 0.867744
[1642]	valid_0's multi_logloss: 0.867957
[1643]	valid_0's multi_logloss: 0.86812
[1644]	valid_0's multi_logloss: 0.868435
[1645]	valid_0's multi_logloss: 0.868641
[1646]	valid_0's multi_logloss: 0.868831
[1647]	valid_0's multi_logloss: 0.8691
[1648]	valid_0's multi_logloss: 0.869369
[1649]	valid_0's multi_logloss: 0.869623
[1650]	valid_0's multi_logloss: 0.869877
[1651]	valid_0's multi_logloss: 0.870113
[1652]	valid_0's multi_logloss: 0.870373
[1653]	valid_0's multi_logloss: 0.870613
[1654]	valid_0's multi_logloss: 0.870988
[1655]	valid_0's multi_logloss: 0.871236
[1656]	valid_0's multi_logloss: 0.871433
[1657]	valid_0's multi_logloss: 0.871644
[1658]	valid_0's multi_logloss: 0.871993
[1659]	valid_0's mult

[1836]	valid_0's multi_logloss: 0.914077
[1837]	valid_0's multi_logloss: 0.914317
[1838]	valid_0's multi_logloss: 0.914578
[1839]	valid_0's multi_logloss: 0.914928
[1840]	valid_0's multi_logloss: 0.91525
[1841]	valid_0's multi_logloss: 0.915425
[1842]	valid_0's multi_logloss: 0.915718
[1843]	valid_0's multi_logloss: 0.915939
[1844]	valid_0's multi_logloss: 0.916287
[1845]	valid_0's multi_logloss: 0.916496
[1846]	valid_0's multi_logloss: 0.916689
[1847]	valid_0's multi_logloss: 0.91699
[1848]	valid_0's multi_logloss: 0.91727
[1849]	valid_0's multi_logloss: 0.91759
[1850]	valid_0's multi_logloss: 0.917992
[1851]	valid_0's multi_logloss: 0.918196
[1852]	valid_0's multi_logloss: 0.91844
[1853]	valid_0's multi_logloss: 0.91861
[1854]	valid_0's multi_logloss: 0.918853
[1855]	valid_0's multi_logloss: 0.919149
[1856]	valid_0's multi_logloss: 0.919378
[1857]	valid_0's multi_logloss: 0.919554
[1858]	valid_0's multi_logloss: 0.919718
[1859]	valid_0's multi_logloss: 0.919908
[1860]	valid_0's multi

[39]	valid_0's multi_logloss: 0.908666
[40]	valid_0's multi_logloss: 0.902846
[41]	valid_0's multi_logloss: 0.897212
[42]	valid_0's multi_logloss: 0.891353
[43]	valid_0's multi_logloss: 0.885244
[44]	valid_0's multi_logloss: 0.880346
[45]	valid_0's multi_logloss: 0.87535
[46]	valid_0's multi_logloss: 0.870419
[47]	valid_0's multi_logloss: 0.865279
[48]	valid_0's multi_logloss: 0.86061
[49]	valid_0's multi_logloss: 0.855295
[50]	valid_0's multi_logloss: 0.850611
[51]	valid_0's multi_logloss: 0.845521
[52]	valid_0's multi_logloss: 0.841174
[53]	valid_0's multi_logloss: 0.836582
[54]	valid_0's multi_logloss: 0.832516
[55]	valid_0's multi_logloss: 0.827945
[56]	valid_0's multi_logloss: 0.824093
[57]	valid_0's multi_logloss: 0.820142
[58]	valid_0's multi_logloss: 0.816031
[59]	valid_0's multi_logloss: 0.812201
[60]	valid_0's multi_logloss: 0.808214
[61]	valid_0's multi_logloss: 0.804408
[62]	valid_0's multi_logloss: 0.801028
[63]	valid_0's multi_logloss: 0.797808
[64]	valid_0's multi_loglos

[246]	valid_0's multi_logloss: 0.627901
[247]	valid_0's multi_logloss: 0.627886
[248]	valid_0's multi_logloss: 0.627627
[249]	valid_0's multi_logloss: 0.627506
[250]	valid_0's multi_logloss: 0.627314
[251]	valid_0's multi_logloss: 0.627195
[252]	valid_0's multi_logloss: 0.62716
[253]	valid_0's multi_logloss: 0.62707
[254]	valid_0's multi_logloss: 0.626994
[255]	valid_0's multi_logloss: 0.626901
[256]	valid_0's multi_logloss: 0.626822
[257]	valid_0's multi_logloss: 0.626725
[258]	valid_0's multi_logloss: 0.626546
[259]	valid_0's multi_logloss: 0.626625
[260]	valid_0's multi_logloss: 0.626426
[261]	valid_0's multi_logloss: 0.62637
[262]	valid_0's multi_logloss: 0.626227
[263]	valid_0's multi_logloss: 0.626305
[264]	valid_0's multi_logloss: 0.626185
[265]	valid_0's multi_logloss: 0.626303
[266]	valid_0's multi_logloss: 0.626072
[267]	valid_0's multi_logloss: 0.626023
[268]	valid_0's multi_logloss: 0.625977
[269]	valid_0's multi_logloss: 0.625903
[270]	valid_0's multi_logloss: 0.625915
[27

[452]	valid_0's multi_logloss: 0.633077
[453]	valid_0's multi_logloss: 0.633241
[454]	valid_0's multi_logloss: 0.633429
[455]	valid_0's multi_logloss: 0.633507
[456]	valid_0's multi_logloss: 0.633607
[457]	valid_0's multi_logloss: 0.633714
[458]	valid_0's multi_logloss: 0.633904
[459]	valid_0's multi_logloss: 0.634014
[460]	valid_0's multi_logloss: 0.63409
[461]	valid_0's multi_logloss: 0.634258
[462]	valid_0's multi_logloss: 0.634349
[463]	valid_0's multi_logloss: 0.634551
[464]	valid_0's multi_logloss: 0.634752
[465]	valid_0's multi_logloss: 0.634972
[466]	valid_0's multi_logloss: 0.635028
[467]	valid_0's multi_logloss: 0.635216
[468]	valid_0's multi_logloss: 0.635333
[469]	valid_0's multi_logloss: 0.635453
[470]	valid_0's multi_logloss: 0.63552
[471]	valid_0's multi_logloss: 0.635593
[472]	valid_0's multi_logloss: 0.635705
[473]	valid_0's multi_logloss: 0.635915
[474]	valid_0's multi_logloss: 0.636091
[475]	valid_0's multi_logloss: 0.636196
[476]	valid_0's multi_logloss: 0.636302
[4

[658]	valid_0's multi_logloss: 0.666148
[659]	valid_0's multi_logloss: 0.66633
[660]	valid_0's multi_logloss: 0.666547
[661]	valid_0's multi_logloss: 0.666772
[662]	valid_0's multi_logloss: 0.666917
[663]	valid_0's multi_logloss: 0.667221
[664]	valid_0's multi_logloss: 0.667422
[665]	valid_0's multi_logloss: 0.667601
[666]	valid_0's multi_logloss: 0.667792
[667]	valid_0's multi_logloss: 0.667918
[668]	valid_0's multi_logloss: 0.668105
[669]	valid_0's multi_logloss: 0.668283
[670]	valid_0's multi_logloss: 0.668483
[671]	valid_0's multi_logloss: 0.668683
[672]	valid_0's multi_logloss: 0.668843
[673]	valid_0's multi_logloss: 0.669026
[674]	valid_0's multi_logloss: 0.669147
[675]	valid_0's multi_logloss: 0.669285
[676]	valid_0's multi_logloss: 0.669422
[677]	valid_0's multi_logloss: 0.669677
[678]	valid_0's multi_logloss: 0.669845
[679]	valid_0's multi_logloss: 0.670132
[680]	valid_0's multi_logloss: 0.670211
[681]	valid_0's multi_logloss: 0.670377
[682]	valid_0's multi_logloss: 0.670509
[

[864]	valid_0's multi_logloss: 0.709557
[865]	valid_0's multi_logloss: 0.709789
[866]	valid_0's multi_logloss: 0.709972
[867]	valid_0's multi_logloss: 0.710218
[868]	valid_0's multi_logloss: 0.710509
[869]	valid_0's multi_logloss: 0.710814
[870]	valid_0's multi_logloss: 0.711049
[871]	valid_0's multi_logloss: 0.711287
[872]	valid_0's multi_logloss: 0.711499
[873]	valid_0's multi_logloss: 0.711748
[874]	valid_0's multi_logloss: 0.711932
[875]	valid_0's multi_logloss: 0.712143
[876]	valid_0's multi_logloss: 0.71237
[877]	valid_0's multi_logloss: 0.712661
[878]	valid_0's multi_logloss: 0.712886
[879]	valid_0's multi_logloss: 0.713169
[880]	valid_0's multi_logloss: 0.713317
[881]	valid_0's multi_logloss: 0.713523
[882]	valid_0's multi_logloss: 0.713605
[883]	valid_0's multi_logloss: 0.713781
[884]	valid_0's multi_logloss: 0.713932
[885]	valid_0's multi_logloss: 0.714225
[886]	valid_0's multi_logloss: 0.714482
[887]	valid_0's multi_logloss: 0.714694
[888]	valid_0's multi_logloss: 0.715039
[

[1068]	valid_0's multi_logloss: 0.756097
[1069]	valid_0's multi_logloss: 0.756313
[1070]	valid_0's multi_logloss: 0.756518
[1071]	valid_0's multi_logloss: 0.756752
[1072]	valid_0's multi_logloss: 0.756944
[1073]	valid_0's multi_logloss: 0.75723
[1074]	valid_0's multi_logloss: 0.757535
[1075]	valid_0's multi_logloss: 0.75769
[1076]	valid_0's multi_logloss: 0.757881
[1077]	valid_0's multi_logloss: 0.758089
[1078]	valid_0's multi_logloss: 0.758337
[1079]	valid_0's multi_logloss: 0.758528
[1080]	valid_0's multi_logloss: 0.758742
[1081]	valid_0's multi_logloss: 0.758922
[1082]	valid_0's multi_logloss: 0.759233
[1083]	valid_0's multi_logloss: 0.759523
[1084]	valid_0's multi_logloss: 0.759849
[1085]	valid_0's multi_logloss: 0.760155
[1086]	valid_0's multi_logloss: 0.760377
[1087]	valid_0's multi_logloss: 0.760596
[1088]	valid_0's multi_logloss: 0.760773
[1089]	valid_0's multi_logloss: 0.760979
[1090]	valid_0's multi_logloss: 0.761214
[1091]	valid_0's multi_logloss: 0.761424
[1092]	valid_0's m

[1269]	valid_0's multi_logloss: 0.804783
[1270]	valid_0's multi_logloss: 0.804997
[1271]	valid_0's multi_logloss: 0.80522
[1272]	valid_0's multi_logloss: 0.805481
[1273]	valid_0's multi_logloss: 0.80571
[1274]	valid_0's multi_logloss: 0.805954
[1275]	valid_0's multi_logloss: 0.806201
[1276]	valid_0's multi_logloss: 0.806395
[1277]	valid_0's multi_logloss: 0.806633
[1278]	valid_0's multi_logloss: 0.80696
[1279]	valid_0's multi_logloss: 0.807197
[1280]	valid_0's multi_logloss: 0.807537
[1281]	valid_0's multi_logloss: 0.807826
[1282]	valid_0's multi_logloss: 0.808078
[1283]	valid_0's multi_logloss: 0.808217
[1284]	valid_0's multi_logloss: 0.808485
[1285]	valid_0's multi_logloss: 0.808764
[1286]	valid_0's multi_logloss: 0.809035
[1287]	valid_0's multi_logloss: 0.8092
[1288]	valid_0's multi_logloss: 0.809495
[1289]	valid_0's multi_logloss: 0.809861
[1290]	valid_0's multi_logloss: 0.810108
[1291]	valid_0's multi_logloss: 0.810392
[1292]	valid_0's multi_logloss: 0.810637
[1293]	valid_0's mult

[1470]	valid_0's multi_logloss: 0.85549
[1471]	valid_0's multi_logloss: 0.855779
[1472]	valid_0's multi_logloss: 0.856099
[1473]	valid_0's multi_logloss: 0.856328
[1474]	valid_0's multi_logloss: 0.856628
[1475]	valid_0's multi_logloss: 0.856913
[1476]	valid_0's multi_logloss: 0.857169
[1477]	valid_0's multi_logloss: 0.857442
[1478]	valid_0's multi_logloss: 0.857695
[1479]	valid_0's multi_logloss: 0.857954
[1480]	valid_0's multi_logloss: 0.858222
[1481]	valid_0's multi_logloss: 0.858527
[1482]	valid_0's multi_logloss: 0.858765
[1483]	valid_0's multi_logloss: 0.858939
[1484]	valid_0's multi_logloss: 0.859216
[1485]	valid_0's multi_logloss: 0.859534
[1486]	valid_0's multi_logloss: 0.859769
[1487]	valid_0's multi_logloss: 0.86004
[1488]	valid_0's multi_logloss: 0.860316
[1489]	valid_0's multi_logloss: 0.860578
[1490]	valid_0's multi_logloss: 0.86073
[1491]	valid_0's multi_logloss: 0.861014
[1492]	valid_0's multi_logloss: 0.861339
[1493]	valid_0's multi_logloss: 0.861656
[1494]	valid_0's mu

[1671]	valid_0's multi_logloss: 0.907411
[1672]	valid_0's multi_logloss: 0.907644
[1673]	valid_0's multi_logloss: 0.907892
[1674]	valid_0's multi_logloss: 0.908103
[1675]	valid_0's multi_logloss: 0.908427
[1676]	valid_0's multi_logloss: 0.908764
[1677]	valid_0's multi_logloss: 0.908881
[1678]	valid_0's multi_logloss: 0.909096
[1679]	valid_0's multi_logloss: 0.90935
[1680]	valid_0's multi_logloss: 0.909672
[1681]	valid_0's multi_logloss: 0.909844
[1682]	valid_0's multi_logloss: 0.910091
[1683]	valid_0's multi_logloss: 0.910454
[1684]	valid_0's multi_logloss: 0.910718
[1685]	valid_0's multi_logloss: 0.910943
[1686]	valid_0's multi_logloss: 0.91112
[1687]	valid_0's multi_logloss: 0.911342
[1688]	valid_0's multi_logloss: 0.911539
[1689]	valid_0's multi_logloss: 0.911767
[1690]	valid_0's multi_logloss: 0.912027
[1691]	valid_0's multi_logloss: 0.912254
[1692]	valid_0's multi_logloss: 0.912624
[1693]	valid_0's multi_logloss: 0.913046
[1694]	valid_0's multi_logloss: 0.913295
[1695]	valid_0's m

[1872]	valid_0's multi_logloss: 0.959296
[1873]	valid_0's multi_logloss: 0.95954
[1874]	valid_0's multi_logloss: 0.95982
[1875]	valid_0's multi_logloss: 0.960006
[1876]	valid_0's multi_logloss: 0.960264
[1877]	valid_0's multi_logloss: 0.960532
[1878]	valid_0's multi_logloss: 0.960763
[1879]	valid_0's multi_logloss: 0.960915
[1880]	valid_0's multi_logloss: 0.961143
[1881]	valid_0's multi_logloss: 0.961389
[1882]	valid_0's multi_logloss: 0.961654
[1883]	valid_0's multi_logloss: 0.96177
[1884]	valid_0's multi_logloss: 0.961987
[1885]	valid_0's multi_logloss: 0.962269
[1886]	valid_0's multi_logloss: 0.962511
[1887]	valid_0's multi_logloss: 0.962733
[1888]	valid_0's multi_logloss: 0.962897
[1889]	valid_0's multi_logloss: 0.963199
[1890]	valid_0's multi_logloss: 0.963484
[1891]	valid_0's multi_logloss: 0.963753
[1892]	valid_0's multi_logloss: 0.964009
[1893]	valid_0's multi_logloss: 0.964313
[1894]	valid_0's multi_logloss: 0.964587
[1895]	valid_0's multi_logloss: 0.964722
[1896]	valid_0's mu

[77]	valid_0's multi_logloss: 0.748565
[78]	valid_0's multi_logloss: 0.746217
[79]	valid_0's multi_logloss: 0.743735
[80]	valid_0's multi_logloss: 0.74139
[81]	valid_0's multi_logloss: 0.738936
[82]	valid_0's multi_logloss: 0.736899
[83]	valid_0's multi_logloss: 0.734591
[84]	valid_0's multi_logloss: 0.732226
[85]	valid_0's multi_logloss: 0.730026
[86]	valid_0's multi_logloss: 0.72808
[87]	valid_0's multi_logloss: 0.726174
[88]	valid_0's multi_logloss: 0.723989
[89]	valid_0's multi_logloss: 0.722009
[90]	valid_0's multi_logloss: 0.72012
[91]	valid_0's multi_logloss: 0.718096
[92]	valid_0's multi_logloss: 0.716371
[93]	valid_0's multi_logloss: 0.71436
[94]	valid_0's multi_logloss: 0.712374
[95]	valid_0's multi_logloss: 0.710764
[96]	valid_0's multi_logloss: 0.709267
[97]	valid_0's multi_logloss: 0.707617
[98]	valid_0's multi_logloss: 0.706026
[99]	valid_0's multi_logloss: 0.704469
[100]	valid_0's multi_logloss: 0.702939
[101]	valid_0's multi_logloss: 0.701486
[102]	valid_0's multi_loglo

[283]	valid_0's multi_logloss: 0.624004
[284]	valid_0's multi_logloss: 0.623894
[285]	valid_0's multi_logloss: 0.623859
[286]	valid_0's multi_logloss: 0.6239
[287]	valid_0's multi_logloss: 0.623985
[288]	valid_0's multi_logloss: 0.62392
[289]	valid_0's multi_logloss: 0.623877
[290]	valid_0's multi_logloss: 0.623807
[291]	valid_0's multi_logloss: 0.623833
[292]	valid_0's multi_logloss: 0.62395
[293]	valid_0's multi_logloss: 0.624052
[294]	valid_0's multi_logloss: 0.624077
[295]	valid_0's multi_logloss: 0.624136
[296]	valid_0's multi_logloss: 0.624169
[297]	valid_0's multi_logloss: 0.624213
[298]	valid_0's multi_logloss: 0.624219
[299]	valid_0's multi_logloss: 0.62427
[300]	valid_0's multi_logloss: 0.624291
[301]	valid_0's multi_logloss: 0.624389
[302]	valid_0's multi_logloss: 0.624367
[303]	valid_0's multi_logloss: 0.624528
[304]	valid_0's multi_logloss: 0.624496
[305]	valid_0's multi_logloss: 0.624526
[306]	valid_0's multi_logloss: 0.624578
[307]	valid_0's multi_logloss: 0.62472
[308]	

[489]	valid_0's multi_logloss: 0.642641
[490]	valid_0's multi_logloss: 0.642803
[491]	valid_0's multi_logloss: 0.643005
[492]	valid_0's multi_logloss: 0.643145
[493]	valid_0's multi_logloss: 0.643401
[494]	valid_0's multi_logloss: 0.643511
[495]	valid_0's multi_logloss: 0.643706
[496]	valid_0's multi_logloss: 0.643893
[497]	valid_0's multi_logloss: 0.644065
[498]	valid_0's multi_logloss: 0.644226
[499]	valid_0's multi_logloss: 0.644324
[500]	valid_0's multi_logloss: 0.644469
[501]	valid_0's multi_logloss: 0.644641
[502]	valid_0's multi_logloss: 0.644774
[503]	valid_0's multi_logloss: 0.644922
[504]	valid_0's multi_logloss: 0.645089
[505]	valid_0's multi_logloss: 0.645357
[506]	valid_0's multi_logloss: 0.645539
[507]	valid_0's multi_logloss: 0.645599
[508]	valid_0's multi_logloss: 0.645763
[509]	valid_0's multi_logloss: 0.64596
[510]	valid_0's multi_logloss: 0.646198
[511]	valid_0's multi_logloss: 0.64635
[512]	valid_0's multi_logloss: 0.646506
[513]	valid_0's multi_logloss: 0.646687
[5

[695]	valid_0's multi_logloss: 0.680355
[696]	valid_0's multi_logloss: 0.680538
[697]	valid_0's multi_logloss: 0.680656
[698]	valid_0's multi_logloss: 0.680863
[699]	valid_0's multi_logloss: 0.681075
[700]	valid_0's multi_logloss: 0.681174
[701]	valid_0's multi_logloss: 0.681402
[702]	valid_0's multi_logloss: 0.681501
[703]	valid_0's multi_logloss: 0.681691
[704]	valid_0's multi_logloss: 0.681881
[705]	valid_0's multi_logloss: 0.682004
[706]	valid_0's multi_logloss: 0.682131
[707]	valid_0's multi_logloss: 0.682331
[708]	valid_0's multi_logloss: 0.682511
[709]	valid_0's multi_logloss: 0.682707
[710]	valid_0's multi_logloss: 0.682877
[711]	valid_0's multi_logloss: 0.6831
[712]	valid_0's multi_logloss: 0.683374
[713]	valid_0's multi_logloss: 0.683522
[714]	valid_0's multi_logloss: 0.683613
[715]	valid_0's multi_logloss: 0.683845
[716]	valid_0's multi_logloss: 0.684025
[717]	valid_0's multi_logloss: 0.68423
[718]	valid_0's multi_logloss: 0.684562
[719]	valid_0's multi_logloss: 0.684707
[72

[901]	valid_0's multi_logloss: 0.724051
[902]	valid_0's multi_logloss: 0.724261
[903]	valid_0's multi_logloss: 0.7244
[904]	valid_0's multi_logloss: 0.724662
[905]	valid_0's multi_logloss: 0.724892
[906]	valid_0's multi_logloss: 0.725187
[907]	valid_0's multi_logloss: 0.725454
[908]	valid_0's multi_logloss: 0.72556
[909]	valid_0's multi_logloss: 0.725724
[910]	valid_0's multi_logloss: 0.725952
[911]	valid_0's multi_logloss: 0.726156
[912]	valid_0's multi_logloss: 0.726398
[913]	valid_0's multi_logloss: 0.726587
[914]	valid_0's multi_logloss: 0.726758
[915]	valid_0's multi_logloss: 0.726947
[916]	valid_0's multi_logloss: 0.727169
[917]	valid_0's multi_logloss: 0.727338
[918]	valid_0's multi_logloss: 0.7275
[919]	valid_0's multi_logloss: 0.727889
[920]	valid_0's multi_logloss: 0.72813
[921]	valid_0's multi_logloss: 0.728399
[922]	valid_0's multi_logloss: 0.72851
[923]	valid_0's multi_logloss: 0.728841
[924]	valid_0's multi_logloss: 0.72898
[925]	valid_0's multi_logloss: 0.729255
[926]	va

[1104]	valid_0's multi_logloss: 0.770897
[1105]	valid_0's multi_logloss: 0.771038
[1106]	valid_0's multi_logloss: 0.771303
[1107]	valid_0's multi_logloss: 0.771491
[1108]	valid_0's multi_logloss: 0.77178
[1109]	valid_0's multi_logloss: 0.772068
[1110]	valid_0's multi_logloss: 0.772321
[1111]	valid_0's multi_logloss: 0.772593
[1112]	valid_0's multi_logloss: 0.772765
[1113]	valid_0's multi_logloss: 0.772953
[1114]	valid_0's multi_logloss: 0.773183
[1115]	valid_0's multi_logloss: 0.773416
[1116]	valid_0's multi_logloss: 0.773666
[1117]	valid_0's multi_logloss: 0.774
[1118]	valid_0's multi_logloss: 0.774299
[1119]	valid_0's multi_logloss: 0.774488
[1120]	valid_0's multi_logloss: 0.774734
[1121]	valid_0's multi_logloss: 0.774981
[1122]	valid_0's multi_logloss: 0.775161
[1123]	valid_0's multi_logloss: 0.775434
[1124]	valid_0's multi_logloss: 0.775718
[1125]	valid_0's multi_logloss: 0.776034
[1126]	valid_0's multi_logloss: 0.776229
[1127]	valid_0's multi_logloss: 0.776491
[1128]	valid_0's mul

[1305]	valid_0's multi_logloss: 0.819741
[1306]	valid_0's multi_logloss: 0.820006
[1307]	valid_0's multi_logloss: 0.820206
[1308]	valid_0's multi_logloss: 0.820409
[1309]	valid_0's multi_logloss: 0.82059
[1310]	valid_0's multi_logloss: 0.8208
[1311]	valid_0's multi_logloss: 0.821011
[1312]	valid_0's multi_logloss: 0.821368
[1313]	valid_0's multi_logloss: 0.821632
[1314]	valid_0's multi_logloss: 0.821843
[1315]	valid_0's multi_logloss: 0.822013
[1316]	valid_0's multi_logloss: 0.822214
[1317]	valid_0's multi_logloss: 0.822559
[1318]	valid_0's multi_logloss: 0.822724
[1319]	valid_0's multi_logloss: 0.822912
[1320]	valid_0's multi_logloss: 0.823146
[1321]	valid_0's multi_logloss: 0.823395
[1322]	valid_0's multi_logloss: 0.823682
[1323]	valid_0's multi_logloss: 0.823951
[1324]	valid_0's multi_logloss: 0.824277
[1325]	valid_0's multi_logloss: 0.824484
[1326]	valid_0's multi_logloss: 0.824687
[1327]	valid_0's multi_logloss: 0.825038
[1328]	valid_0's multi_logloss: 0.825347
[1329]	valid_0's mu

[1506]	valid_0's multi_logloss: 0.87166
[1507]	valid_0's multi_logloss: 0.87194
[1508]	valid_0's multi_logloss: 0.87218
[1509]	valid_0's multi_logloss: 0.872524
[1510]	valid_0's multi_logloss: 0.872765
[1511]	valid_0's multi_logloss: 0.873103
[1512]	valid_0's multi_logloss: 0.873342
[1513]	valid_0's multi_logloss: 0.873641
[1514]	valid_0's multi_logloss: 0.873842
[1515]	valid_0's multi_logloss: 0.874122
[1516]	valid_0's multi_logloss: 0.874407
[1517]	valid_0's multi_logloss: 0.874591
[1518]	valid_0's multi_logloss: 0.874878
[1519]	valid_0's multi_logloss: 0.87509
[1520]	valid_0's multi_logloss: 0.87539
[1521]	valid_0's multi_logloss: 0.875589
[1522]	valid_0's multi_logloss: 0.875944
[1523]	valid_0's multi_logloss: 0.876233
[1524]	valid_0's multi_logloss: 0.876589
[1525]	valid_0's multi_logloss: 0.87677
[1526]	valid_0's multi_logloss: 0.876936
[1527]	valid_0's multi_logloss: 0.87732
[1528]	valid_0's multi_logloss: 0.877611
[1529]	valid_0's multi_logloss: 0.877929
[1530]	valid_0's multi_

[1707]	valid_0's multi_logloss: 0.925241
[1708]	valid_0's multi_logloss: 0.925509
[1709]	valid_0's multi_logloss: 0.925667
[1710]	valid_0's multi_logloss: 0.92594
[1711]	valid_0's multi_logloss: 0.926196
[1712]	valid_0's multi_logloss: 0.92656
[1713]	valid_0's multi_logloss: 0.926859
[1714]	valid_0's multi_logloss: 0.927044
[1715]	valid_0's multi_logloss: 0.927224
[1716]	valid_0's multi_logloss: 0.927402
[1717]	valid_0's multi_logloss: 0.927719
[1718]	valid_0's multi_logloss: 0.927984
[1719]	valid_0's multi_logloss: 0.928312
[1720]	valid_0's multi_logloss: 0.928549
[1721]	valid_0's multi_logloss: 0.928819
[1722]	valid_0's multi_logloss: 0.929071
[1723]	valid_0's multi_logloss: 0.929296
[1724]	valid_0's multi_logloss: 0.929567
[1725]	valid_0's multi_logloss: 0.929769
[1726]	valid_0's multi_logloss: 0.930046
[1727]	valid_0's multi_logloss: 0.930339
[1728]	valid_0's multi_logloss: 0.930502
[1729]	valid_0's multi_logloss: 0.930679
[1730]	valid_0's multi_logloss: 0.930935
[1731]	valid_0's m

[1908]	valid_0's multi_logloss: 0.975548
[1909]	valid_0's multi_logloss: 0.975768
[1910]	valid_0's multi_logloss: 0.976011
[1911]	valid_0's multi_logloss: 0.976261
[1912]	valid_0's multi_logloss: 0.976504
[1913]	valid_0's multi_logloss: 0.976762
[1914]	valid_0's multi_logloss: 0.977068
[1915]	valid_0's multi_logloss: 0.977317
[1916]	valid_0's multi_logloss: 0.97758
[1917]	valid_0's multi_logloss: 0.977859
[1918]	valid_0's multi_logloss: 0.978077
[1919]	valid_0's multi_logloss: 0.978296
[1920]	valid_0's multi_logloss: 0.978571
[1921]	valid_0's multi_logloss: 0.978773
[1922]	valid_0's multi_logloss: 0.978939
[1923]	valid_0's multi_logloss: 0.979279
[1924]	valid_0's multi_logloss: 0.979525
[1925]	valid_0's multi_logloss: 0.979771
[1926]	valid_0's multi_logloss: 0.979963
[1927]	valid_0's multi_logloss: 0.98025
[1928]	valid_0's multi_logloss: 0.980411
[1929]	valid_0's multi_logloss: 0.980658
[1930]	valid_0's multi_logloss: 0.980905
[1931]	valid_0's multi_logloss: 0.98114
[1932]	valid_0's mu

[114]	valid_0's multi_logloss: 0.687083
[115]	valid_0's multi_logloss: 0.685816
[116]	valid_0's multi_logloss: 0.684538
[117]	valid_0's multi_logloss: 0.683429
[118]	valid_0's multi_logloss: 0.682346
[119]	valid_0's multi_logloss: 0.681167
[120]	valid_0's multi_logloss: 0.679881
[121]	valid_0's multi_logloss: 0.678831
[122]	valid_0's multi_logloss: 0.677592
[123]	valid_0's multi_logloss: 0.676576
[124]	valid_0's multi_logloss: 0.675492
[125]	valid_0's multi_logloss: 0.674656
[126]	valid_0's multi_logloss: 0.673657
[127]	valid_0's multi_logloss: 0.672648
[128]	valid_0's multi_logloss: 0.671602
[129]	valid_0's multi_logloss: 0.67061
[130]	valid_0's multi_logloss: 0.669555
[131]	valid_0's multi_logloss: 0.668504
[132]	valid_0's multi_logloss: 0.667375
[133]	valid_0's multi_logloss: 0.666369
[134]	valid_0's multi_logloss: 0.665587
[135]	valid_0's multi_logloss: 0.664783
[136]	valid_0's multi_logloss: 0.664046
[137]	valid_0's multi_logloss: 0.663166
[138]	valid_0's multi_logloss: 0.66234
[1

[320]	valid_0's multi_logloss: 0.620553
[321]	valid_0's multi_logloss: 0.620548
[322]	valid_0's multi_logloss: 0.620658
[323]	valid_0's multi_logloss: 0.620676
[324]	valid_0's multi_logloss: 0.62065
[325]	valid_0's multi_logloss: 0.62067
[326]	valid_0's multi_logloss: 0.620669
[327]	valid_0's multi_logloss: 0.620602
[328]	valid_0's multi_logloss: 0.620518
[329]	valid_0's multi_logloss: 0.620499
[330]	valid_0's multi_logloss: 0.620577
[331]	valid_0's multi_logloss: 0.620659
[332]	valid_0's multi_logloss: 0.620668
[333]	valid_0's multi_logloss: 0.620809
[334]	valid_0's multi_logloss: 0.620765
[335]	valid_0's multi_logloss: 0.620752
[336]	valid_0's multi_logloss: 0.62086
[337]	valid_0's multi_logloss: 0.620891
[338]	valid_0's multi_logloss: 0.62084
[339]	valid_0's multi_logloss: 0.620792
[340]	valid_0's multi_logloss: 0.620834
[341]	valid_0's multi_logloss: 0.62076
[342]	valid_0's multi_logloss: 0.620922
[343]	valid_0's multi_logloss: 0.620982
[344]	valid_0's multi_logloss: 0.620929
[345]

[526]	valid_0's multi_logloss: 0.641767
[527]	valid_0's multi_logloss: 0.64195
[528]	valid_0's multi_logloss: 0.642079
[529]	valid_0's multi_logloss: 0.642223
[530]	valid_0's multi_logloss: 0.642362
[531]	valid_0's multi_logloss: 0.642596
[532]	valid_0's multi_logloss: 0.642754
[533]	valid_0's multi_logloss: 0.64302
[534]	valid_0's multi_logloss: 0.643211
[535]	valid_0's multi_logloss: 0.643389
[536]	valid_0's multi_logloss: 0.643549
[537]	valid_0's multi_logloss: 0.643752
[538]	valid_0's multi_logloss: 0.643895
[539]	valid_0's multi_logloss: 0.644108
[540]	valid_0's multi_logloss: 0.64421
[541]	valid_0's multi_logloss: 0.644322
[542]	valid_0's multi_logloss: 0.644379
[543]	valid_0's multi_logloss: 0.644535
[544]	valid_0's multi_logloss: 0.644724
[545]	valid_0's multi_logloss: 0.644831
[546]	valid_0's multi_logloss: 0.644975
[547]	valid_0's multi_logloss: 0.645299
[548]	valid_0's multi_logloss: 0.645415
[549]	valid_0's multi_logloss: 0.64561
[550]	valid_0's multi_logloss: 0.64587
[551]

[732]	valid_0's multi_logloss: 0.676984
[733]	valid_0's multi_logloss: 0.677058
[734]	valid_0's multi_logloss: 0.677291
[735]	valid_0's multi_logloss: 0.677445
[736]	valid_0's multi_logloss: 0.677629
[737]	valid_0's multi_logloss: 0.677838
[738]	valid_0's multi_logloss: 0.678008
[739]	valid_0's multi_logloss: 0.678207
[740]	valid_0's multi_logloss: 0.678434
[741]	valid_0's multi_logloss: 0.67866
[742]	valid_0's multi_logloss: 0.678813
[743]	valid_0's multi_logloss: 0.678932
[744]	valid_0's multi_logloss: 0.679092
[745]	valid_0's multi_logloss: 0.679333
[746]	valid_0's multi_logloss: 0.679534
[747]	valid_0's multi_logloss: 0.679726
[748]	valid_0's multi_logloss: 0.67988
[749]	valid_0's multi_logloss: 0.680047
[750]	valid_0's multi_logloss: 0.680201
[751]	valid_0's multi_logloss: 0.680379
[752]	valid_0's multi_logloss: 0.680509
[753]	valid_0's multi_logloss: 0.680711
[754]	valid_0's multi_logloss: 0.680956
[755]	valid_0's multi_logloss: 0.681214
[756]	valid_0's multi_logloss: 0.681372
[7

[938]	valid_0's multi_logloss: 0.718545
[939]	valid_0's multi_logloss: 0.718713
[940]	valid_0's multi_logloss: 0.71892
[941]	valid_0's multi_logloss: 0.719142
[942]	valid_0's multi_logloss: 0.71932
[943]	valid_0's multi_logloss: 0.719455
[944]	valid_0's multi_logloss: 0.719702
[945]	valid_0's multi_logloss: 0.719878
[946]	valid_0's multi_logloss: 0.720071
[947]	valid_0's multi_logloss: 0.720134
[948]	valid_0's multi_logloss: 0.720416
[949]	valid_0's multi_logloss: 0.720686
[950]	valid_0's multi_logloss: 0.72091
[951]	valid_0's multi_logloss: 0.721016
[952]	valid_0's multi_logloss: 0.721214
[953]	valid_0's multi_logloss: 0.721316
[954]	valid_0's multi_logloss: 0.721601
[955]	valid_0's multi_logloss: 0.721723
[956]	valid_0's multi_logloss: 0.72192
[957]	valid_0's multi_logloss: 0.722113
[958]	valid_0's multi_logloss: 0.722338
[959]	valid_0's multi_logloss: 0.722584
[960]	valid_0's multi_logloss: 0.722819
[961]	valid_0's multi_logloss: 0.723089
[962]	valid_0's multi_logloss: 0.723305
[963

[1140]	valid_0's multi_logloss: 0.765371
[1141]	valid_0's multi_logloss: 0.765607
[1142]	valid_0's multi_logloss: 0.765903
[1143]	valid_0's multi_logloss: 0.766127
[1144]	valid_0's multi_logloss: 0.766325
[1145]	valid_0's multi_logloss: 0.766516
[1146]	valid_0's multi_logloss: 0.766744
[1147]	valid_0's multi_logloss: 0.766988
[1148]	valid_0's multi_logloss: 0.767208
[1149]	valid_0's multi_logloss: 0.767545
[1150]	valid_0's multi_logloss: 0.767751
[1151]	valid_0's multi_logloss: 0.768051
[1152]	valid_0's multi_logloss: 0.768292
[1153]	valid_0's multi_logloss: 0.768507
[1154]	valid_0's multi_logloss: 0.76874
[1155]	valid_0's multi_logloss: 0.768985
[1156]	valid_0's multi_logloss: 0.769318
[1157]	valid_0's multi_logloss: 0.769554
[1158]	valid_0's multi_logloss: 0.769794
[1159]	valid_0's multi_logloss: 0.770015
[1160]	valid_0's multi_logloss: 0.770206
[1161]	valid_0's multi_logloss: 0.77053
[1162]	valid_0's multi_logloss: 0.770833
[1163]	valid_0's multi_logloss: 0.771082
[1164]	valid_0's m

[1341]	valid_0's multi_logloss: 0.81479
[1342]	valid_0's multi_logloss: 0.814986
[1343]	valid_0's multi_logloss: 0.815196
[1344]	valid_0's multi_logloss: 0.815446
[1345]	valid_0's multi_logloss: 0.815733
[1346]	valid_0's multi_logloss: 0.816131
[1347]	valid_0's multi_logloss: 0.816366
[1348]	valid_0's multi_logloss: 0.816564
[1349]	valid_0's multi_logloss: 0.816781
[1350]	valid_0's multi_logloss: 0.817017
[1351]	valid_0's multi_logloss: 0.81726
[1352]	valid_0's multi_logloss: 0.817538
[1353]	valid_0's multi_logloss: 0.817832
[1354]	valid_0's multi_logloss: 0.81811
[1355]	valid_0's multi_logloss: 0.81826
[1356]	valid_0's multi_logloss: 0.818473
[1357]	valid_0's multi_logloss: 0.818685
[1358]	valid_0's multi_logloss: 0.818921
[1359]	valid_0's multi_logloss: 0.819166
[1360]	valid_0's multi_logloss: 0.819376
[1361]	valid_0's multi_logloss: 0.81957
[1362]	valid_0's multi_logloss: 0.819797
[1363]	valid_0's multi_logloss: 0.820062
[1364]	valid_0's multi_logloss: 0.820295
[1365]	valid_0's mult

[1542]	valid_0's multi_logloss: 0.864742
[1543]	valid_0's multi_logloss: 0.864861
[1544]	valid_0's multi_logloss: 0.865022
[1545]	valid_0's multi_logloss: 0.865267
[1546]	valid_0's multi_logloss: 0.865537
[1547]	valid_0's multi_logloss: 0.865837
[1548]	valid_0's multi_logloss: 0.8661
[1549]	valid_0's multi_logloss: 0.866364
[1550]	valid_0's multi_logloss: 0.866583
[1551]	valid_0's multi_logloss: 0.866923
[1552]	valid_0's multi_logloss: 0.867172
[1553]	valid_0's multi_logloss: 0.86732
[1554]	valid_0's multi_logloss: 0.867512
[1555]	valid_0's multi_logloss: 0.8678
[1556]	valid_0's multi_logloss: 0.868108
[1557]	valid_0's multi_logloss: 0.868359
[1558]	valid_0's multi_logloss: 0.868602
[1559]	valid_0's multi_logloss: 0.868936
[1560]	valid_0's multi_logloss: 0.869185
[1561]	valid_0's multi_logloss: 0.869393
[1562]	valid_0's multi_logloss: 0.869552
[1563]	valid_0's multi_logloss: 0.869762
[1564]	valid_0's multi_logloss: 0.870071
[1565]	valid_0's multi_logloss: 0.870317
[1566]	valid_0's mult

[1743]	valid_0's multi_logloss: 0.914736
[1744]	valid_0's multi_logloss: 0.914958
[1745]	valid_0's multi_logloss: 0.915205
[1746]	valid_0's multi_logloss: 0.915454
[1747]	valid_0's multi_logloss: 0.915742
[1748]	valid_0's multi_logloss: 0.916041
[1749]	valid_0's multi_logloss: 0.91623
[1750]	valid_0's multi_logloss: 0.916604
[1751]	valid_0's multi_logloss: 0.916945
[1752]	valid_0's multi_logloss: 0.91724
[1753]	valid_0's multi_logloss: 0.917541
[1754]	valid_0's multi_logloss: 0.917802
[1755]	valid_0's multi_logloss: 0.917988
[1756]	valid_0's multi_logloss: 0.91808
[1757]	valid_0's multi_logloss: 0.918253
[1758]	valid_0's multi_logloss: 0.918506
[1759]	valid_0's multi_logloss: 0.918768
[1760]	valid_0's multi_logloss: 0.918987
[1761]	valid_0's multi_logloss: 0.91919
[1762]	valid_0's multi_logloss: 0.919443
[1763]	valid_0's multi_logloss: 0.919618
[1764]	valid_0's multi_logloss: 0.919887
[1765]	valid_0's multi_logloss: 0.920158
[1766]	valid_0's multi_logloss: 0.920498
[1767]	valid_0's mul

[1944]	valid_0's multi_logloss: 0.964226
[1945]	valid_0's multi_logloss: 0.964401
[1946]	valid_0's multi_logloss: 0.964595
[1947]	valid_0's multi_logloss: 0.964715
[1948]	valid_0's multi_logloss: 0.964885
[1949]	valid_0's multi_logloss: 0.965053
[1950]	valid_0's multi_logloss: 0.965346
[1951]	valid_0's multi_logloss: 0.965596
[1952]	valid_0's multi_logloss: 0.965771
[1953]	valid_0's multi_logloss: 0.965995
[1954]	valid_0's multi_logloss: 0.966175
[1955]	valid_0's multi_logloss: 0.966409
[1956]	valid_0's multi_logloss: 0.96664
[1957]	valid_0's multi_logloss: 0.966869
[1958]	valid_0's multi_logloss: 0.967214
[1959]	valid_0's multi_logloss: 0.967419
[1960]	valid_0's multi_logloss: 0.967561
[1961]	valid_0's multi_logloss: 0.967884
[1962]	valid_0's multi_logloss: 0.968135
[1963]	valid_0's multi_logloss: 0.968392
[1964]	valid_0's multi_logloss: 0.96865
[1965]	valid_0's multi_logloss: 0.968888
[1966]	valid_0's multi_logloss: 0.969157
[1967]	valid_0's multi_logloss: 0.969441
[1968]	valid_0's m

[151]	valid_0's multi_logloss: 0.642567
[152]	valid_0's multi_logloss: 0.641968
[153]	valid_0's multi_logloss: 0.641248
[154]	valid_0's multi_logloss: 0.640571
[155]	valid_0's multi_logloss: 0.640055
[156]	valid_0's multi_logloss: 0.639458
[157]	valid_0's multi_logloss: 0.639028
[158]	valid_0's multi_logloss: 0.63866
[159]	valid_0's multi_logloss: 0.638163
[160]	valid_0's multi_logloss: 0.637739
[161]	valid_0's multi_logloss: 0.6372
[162]	valid_0's multi_logloss: 0.636704
[163]	valid_0's multi_logloss: 0.636169
[164]	valid_0's multi_logloss: 0.635807
[165]	valid_0's multi_logloss: 0.635309
[166]	valid_0's multi_logloss: 0.634763
[167]	valid_0's multi_logloss: 0.634314
[168]	valid_0's multi_logloss: 0.633866
[169]	valid_0's multi_logloss: 0.633168
[170]	valid_0's multi_logloss: 0.632674
[171]	valid_0's multi_logloss: 0.63217
[172]	valid_0's multi_logloss: 0.631838
[173]	valid_0's multi_logloss: 0.631382
[174]	valid_0's multi_logloss: 0.63084
[175]	valid_0's multi_logloss: 0.630502
[176]

[357]	valid_0's multi_logloss: 0.611777
[358]	valid_0's multi_logloss: 0.611815
[359]	valid_0's multi_logloss: 0.61183
[360]	valid_0's multi_logloss: 0.611817
[361]	valid_0's multi_logloss: 0.611829
[362]	valid_0's multi_logloss: 0.611911
[363]	valid_0's multi_logloss: 0.612014
[364]	valid_0's multi_logloss: 0.612181
[365]	valid_0's multi_logloss: 0.612342
[366]	valid_0's multi_logloss: 0.612412
[367]	valid_0's multi_logloss: 0.612543
[368]	valid_0's multi_logloss: 0.612582
[369]	valid_0's multi_logloss: 0.612616
[370]	valid_0's multi_logloss: 0.612633
[371]	valid_0's multi_logloss: 0.612674
[372]	valid_0's multi_logloss: 0.612716
[373]	valid_0's multi_logloss: 0.612815
[374]	valid_0's multi_logloss: 0.612981
[375]	valid_0's multi_logloss: 0.61306
[376]	valid_0's multi_logloss: 0.613184
[377]	valid_0's multi_logloss: 0.613252
[378]	valid_0's multi_logloss: 0.613338
[379]	valid_0's multi_logloss: 0.613347
[380]	valid_0's multi_logloss: 0.61334
[381]	valid_0's multi_logloss: 0.613401
[38

[563]	valid_0's multi_logloss: 0.635315
[564]	valid_0's multi_logloss: 0.635519
[565]	valid_0's multi_logloss: 0.635811
[566]	valid_0's multi_logloss: 0.636014
[567]	valid_0's multi_logloss: 0.636277
[568]	valid_0's multi_logloss: 0.636354
[569]	valid_0's multi_logloss: 0.636476
[570]	valid_0's multi_logloss: 0.636585
[571]	valid_0's multi_logloss: 0.636792
[572]	valid_0's multi_logloss: 0.636967
[573]	valid_0's multi_logloss: 0.637173
[574]	valid_0's multi_logloss: 0.637285
[575]	valid_0's multi_logloss: 0.637408
[576]	valid_0's multi_logloss: 0.637675
[577]	valid_0's multi_logloss: 0.637846
[578]	valid_0's multi_logloss: 0.637952
[579]	valid_0's multi_logloss: 0.63809
[580]	valid_0's multi_logloss: 0.638237
[581]	valid_0's multi_logloss: 0.638336
[582]	valid_0's multi_logloss: 0.638521
[583]	valid_0's multi_logloss: 0.638791
[584]	valid_0's multi_logloss: 0.639017
[585]	valid_0's multi_logloss: 0.639208
[586]	valid_0's multi_logloss: 0.639384
[587]	valid_0's multi_logloss: 0.639561
[

[769]	valid_0's multi_logloss: 0.673153
[770]	valid_0's multi_logloss: 0.673325
[771]	valid_0's multi_logloss: 0.673591
[772]	valid_0's multi_logloss: 0.67382
[773]	valid_0's multi_logloss: 0.674017
[774]	valid_0's multi_logloss: 0.674329
[775]	valid_0's multi_logloss: 0.67469
[776]	valid_0's multi_logloss: 0.674831
[777]	valid_0's multi_logloss: 0.675022
[778]	valid_0's multi_logloss: 0.675172
[779]	valid_0's multi_logloss: 0.675256
[780]	valid_0's multi_logloss: 0.675352
[781]	valid_0's multi_logloss: 0.675573
[782]	valid_0's multi_logloss: 0.675683
[783]	valid_0's multi_logloss: 0.675865
[784]	valid_0's multi_logloss: 0.676123
[785]	valid_0's multi_logloss: 0.676372
[786]	valid_0's multi_logloss: 0.676577
[787]	valid_0's multi_logloss: 0.676809
[788]	valid_0's multi_logloss: 0.677
[789]	valid_0's multi_logloss: 0.677139
[790]	valid_0's multi_logloss: 0.677337
[791]	valid_0's multi_logloss: 0.677471
[792]	valid_0's multi_logloss: 0.677613
[793]	valid_0's multi_logloss: 0.677841
[794]

[975]	valid_0's multi_logloss: 0.715429
[976]	valid_0's multi_logloss: 0.715618
[977]	valid_0's multi_logloss: 0.715922
[978]	valid_0's multi_logloss: 0.716136
[979]	valid_0's multi_logloss: 0.716371
[980]	valid_0's multi_logloss: 0.716573
[981]	valid_0's multi_logloss: 0.716756
[982]	valid_0's multi_logloss: 0.717029
[983]	valid_0's multi_logloss: 0.717304
[984]	valid_0's multi_logloss: 0.717493
[985]	valid_0's multi_logloss: 0.717789
[986]	valid_0's multi_logloss: 0.717959
[987]	valid_0's multi_logloss: 0.718072
[988]	valid_0's multi_logloss: 0.718292
[989]	valid_0's multi_logloss: 0.718465
[990]	valid_0's multi_logloss: 0.718663
[991]	valid_0's multi_logloss: 0.718856
[992]	valid_0's multi_logloss: 0.719161
[993]	valid_0's multi_logloss: 0.719431
[994]	valid_0's multi_logloss: 0.719631
[995]	valid_0's multi_logloss: 0.719778
[996]	valid_0's multi_logloss: 0.719991
[997]	valid_0's multi_logloss: 0.720214
[998]	valid_0's multi_logloss: 0.720409
[999]	valid_0's multi_logloss: 0.720647


[1176]	valid_0's multi_logloss: 0.760666
[1177]	valid_0's multi_logloss: 0.760944
[1178]	valid_0's multi_logloss: 0.761196
[1179]	valid_0's multi_logloss: 0.761332
[1180]	valid_0's multi_logloss: 0.761597
[1181]	valid_0's multi_logloss: 0.761813
[1182]	valid_0's multi_logloss: 0.762093
[1183]	valid_0's multi_logloss: 0.762381
[1184]	valid_0's multi_logloss: 0.762516
[1185]	valid_0's multi_logloss: 0.762685
[1186]	valid_0's multi_logloss: 0.76293
[1187]	valid_0's multi_logloss: 0.763176
[1188]	valid_0's multi_logloss: 0.763336
[1189]	valid_0's multi_logloss: 0.763612
[1190]	valid_0's multi_logloss: 0.763912
[1191]	valid_0's multi_logloss: 0.764059
[1192]	valid_0's multi_logloss: 0.764356
[1193]	valid_0's multi_logloss: 0.764494
[1194]	valid_0's multi_logloss: 0.764811
[1195]	valid_0's multi_logloss: 0.765044
[1196]	valid_0's multi_logloss: 0.765239
[1197]	valid_0's multi_logloss: 0.765372
[1198]	valid_0's multi_logloss: 0.765596
[1199]	valid_0's multi_logloss: 0.765869
[1200]	valid_0's 

[1377]	valid_0's multi_logloss: 0.80677
[1378]	valid_0's multi_logloss: 0.806956
[1379]	valid_0's multi_logloss: 0.807124
[1380]	valid_0's multi_logloss: 0.807336
[1381]	valid_0's multi_logloss: 0.807548
[1382]	valid_0's multi_logloss: 0.807849
[1383]	valid_0's multi_logloss: 0.808077
[1384]	valid_0's multi_logloss: 0.808291
[1385]	valid_0's multi_logloss: 0.808631
[1386]	valid_0's multi_logloss: 0.80884
[1387]	valid_0's multi_logloss: 0.809104
[1388]	valid_0's multi_logloss: 0.809422
[1389]	valid_0's multi_logloss: 0.809634
[1390]	valid_0's multi_logloss: 0.809891
[1391]	valid_0's multi_logloss: 0.810272
[1392]	valid_0's multi_logloss: 0.810475
[1393]	valid_0's multi_logloss: 0.810673
[1394]	valid_0's multi_logloss: 0.810826
[1395]	valid_0's multi_logloss: 0.811115
[1396]	valid_0's multi_logloss: 0.811417
[1397]	valid_0's multi_logloss: 0.811685
[1398]	valid_0's multi_logloss: 0.81188
[1399]	valid_0's multi_logloss: 0.812133
[1400]	valid_0's multi_logloss: 0.812423
[1401]	valid_0's mu

[1578]	valid_0's multi_logloss: 0.855907
[1579]	valid_0's multi_logloss: 0.856185
[1580]	valid_0's multi_logloss: 0.856336
[1581]	valid_0's multi_logloss: 0.856585
[1582]	valid_0's multi_logloss: 0.85676
[1583]	valid_0's multi_logloss: 0.857046
[1584]	valid_0's multi_logloss: 0.857259
[1585]	valid_0's multi_logloss: 0.857556
[1586]	valid_0's multi_logloss: 0.857779
[1587]	valid_0's multi_logloss: 0.85804
[1588]	valid_0's multi_logloss: 0.858235
[1589]	valid_0's multi_logloss: 0.858437
[1590]	valid_0's multi_logloss: 0.858611
[1591]	valid_0's multi_logloss: 0.858861
[1592]	valid_0's multi_logloss: 0.859157
[1593]	valid_0's multi_logloss: 0.859449
[1594]	valid_0's multi_logloss: 0.859694
[1595]	valid_0's multi_logloss: 0.859931
[1596]	valid_0's multi_logloss: 0.860178
[1597]	valid_0's multi_logloss: 0.860511
[1598]	valid_0's multi_logloss: 0.860761
[1599]	valid_0's multi_logloss: 0.860958
[1600]	valid_0's multi_logloss: 0.861171
[1601]	valid_0's multi_logloss: 0.861397
[1602]	valid_0's m

[1779]	valid_0's multi_logloss: 0.905696
[1780]	valid_0's multi_logloss: 0.905936
[1781]	valid_0's multi_logloss: 0.90613
[1782]	valid_0's multi_logloss: 0.906402
[1783]	valid_0's multi_logloss: 0.906656
[1784]	valid_0's multi_logloss: 0.906926
[1785]	valid_0's multi_logloss: 0.907142
[1786]	valid_0's multi_logloss: 0.907233
[1787]	valid_0's multi_logloss: 0.907458
[1788]	valid_0's multi_logloss: 0.907728
[1789]	valid_0's multi_logloss: 0.907912
[1790]	valid_0's multi_logloss: 0.908172
[1791]	valid_0's multi_logloss: 0.908375
[1792]	valid_0's multi_logloss: 0.908727
[1793]	valid_0's multi_logloss: 0.909024
[1794]	valid_0's multi_logloss: 0.909254
[1795]	valid_0's multi_logloss: 0.909517
[1796]	valid_0's multi_logloss: 0.909796
[1797]	valid_0's multi_logloss: 0.910113
[1798]	valid_0's multi_logloss: 0.910305
[1799]	valid_0's multi_logloss: 0.910522
[1800]	valid_0's multi_logloss: 0.910752
[1801]	valid_0's multi_logloss: 0.910951
[1802]	valid_0's multi_logloss: 0.911161
[1803]	valid_0's 

[1980]	valid_0's multi_logloss: 0.95455
[1981]	valid_0's multi_logloss: 0.954803
[1982]	valid_0's multi_logloss: 0.95499
[1983]	valid_0's multi_logloss: 0.955224
[1984]	valid_0's multi_logloss: 0.955488
[1985]	valid_0's multi_logloss: 0.955834
[1986]	valid_0's multi_logloss: 0.956152
[1987]	valid_0's multi_logloss: 0.95628
[1988]	valid_0's multi_logloss: 0.956496
[1989]	valid_0's multi_logloss: 0.956741
[1990]	valid_0's multi_logloss: 0.956915
[1991]	valid_0's multi_logloss: 0.957213
[1992]	valid_0's multi_logloss: 0.957486
[1993]	valid_0's multi_logloss: 0.957773
[1994]	valid_0's multi_logloss: 0.95801
[1995]	valid_0's multi_logloss: 0.958228
[1996]	valid_0's multi_logloss: 0.95858
[1997]	valid_0's multi_logloss: 0.958785
[1998]	valid_0's multi_logloss: 0.95903
[1999]	valid_0's multi_logloss: 0.959222
[2000]	valid_0's multi_logloss: 0.959543


In [32]:
print(f'Accuracy (CV): {accuracy_score(y, np.argmax(p, axis=1)) * 100:8.4f}%')
print(f'Log Loss (CV): {log_loss(pd.get_dummies(y), p):8.4f}')

Accuracy (CV):  76.0947%
Log Loss (CV):   0.9759


#Accuracy (CV):  76.9274%  
#Log Loss (CV):   0.6212  ...contest : 0.4099    
lgb.LGBMClassifier(objective='multiclass', n_estimators=3000,   num_leaves=256,   learning_rate=0.009,  subsample=.8,  subsample_freq=1, num_iterations=2000,
colsample_bytree=.6,   random_state=42,       n_jobs=-1)  
                             
#Accuracy (CV):  76.0947%  
#Log Loss (CV):   0.9759...contest :  0.4318  
lgb.LGBMClassifier(objective='multiclass',
                             n_estimators=1000,
                             num_leaves=127,
                             learning_rate=0.04147603536728929,
                             subsample=.9,
                             subsample_freq=1,
                             num_iterations=2000,
                             colsample_bytree=.8,
                             random_state=42,
                             n_jobs=-1)
                             
                           

## 파일제출

In [33]:
# submission
sample_submission[['0','1','2','3','4']] = p_tst
sample_submission

index         0         1             2             3             4
0          0  0.000509  0.980555  1.852706e-02  4.064327e-04  2.637948e-06
1          1  0.000430  0.998412  5.333983e-05  8.086088e-05  1.023672e-03
2          2  0.999296  0.000016  1.568217e-04  8.760408e-05  4.438005e-04
3          3  0.000003  0.000015  9.999747e-01  5.950140e-06  1.967738e-06
4          4  0.723953  0.057970  1.467240e-01  6.335313e-02  8.000149e-03
...      ...       ...       ...           ...           ...           ...
19612  19612  0.000029  0.999970  7.696671e-10  2.435033e-07  9.413262e-08
19613  19613  0.030582  0.000007  2.349613e-04  1.622589e-05  9.691605e-01
19614  19614  0.000007  0.999992  4.403896e-10  1.034265e-06  3.914624e-09
19615  19615  0.000159  0.999549  5.806682e-05  2.336879e-04  8.621616e-07
19616  19616  0.739638  0.000339  1.676097e-02  1.803189e-01  6.294318e-02

[19617 rows x 6 columns]

In [34]:
sample_submission.to_csv('thisissubmission.csv', index = False, encoding = 'utf-8')

## test 데이터 셋 이용하기
(대회 규정상 이를 제출하면 안됨!!)

In [15]:
df = pd.concat([train, test], axis=0)

In [16]:
print(df.shape)
df.tail()

(74496, 3)


index                                               text  author
19612  19612  At the end of another day or two, odin growing...     NaN
19613  19613  All afternoon we sat together, mostly in silen...     NaN
19614  19614   odin, having carried his thanks to odin, proc...     NaN
19615  19615  Soon after this, upon odin's leaving the room,...     NaN
19616  19616  And all the worse for the doomed man, that the...     NaN

In [17]:
df.fillna(-1, inplace=True)
df.tail()

index                                               text  author
19612  19612  At the end of another day or two, odin growing...    -1.0
19613  19613  All afternoon we sat together, mostly in silen...    -1.0
19614  19614   odin, having carried his thanks to odin, proc...    -1.0
19615  19615  Soon after this, upon odin's leaving the room,...    -1.0
19616  19616  And all the worse for the doomed man, that the...    -1.0

In [18]:
#부호를 제거해주는 함수
def alpha_num(text):
    return re.sub(r'[^A-Za-z0-9 ]', '', text)

df['text']=df['text'].apply(alpha_num)

In [19]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\USER\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [20]:
df=df['text'].values

In [21]:
for i, document in enumerate(df):
    words = word_tokenize(document)
    #print(words) #
    clean_words = []
    for word in words:
        word = word.lower()                         #대소분자 변환
        if word not in stopwords.words('english'): #불용어 제거
            stemmer = SnowballStemmer('english')
            word = stemmer.stem(word)              #어간 추출
            clean_words.append(word)
    document = ' '.join(clean_words)
    df[i] = document

KeyboardInterrupt: 

In [ ]:
df = CountVectorizer(tokenizer=word_tokenize, stop_words='english', min_df=50).fit_transform(df)

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline

vectorizer = CountVectorizer(analyzer = 'word', 
                             tokenizer = None,
                             preprocessor = None, 
                             stop_words = None, 
                             min_df = 2, # 토큰이 나타날 최소 문서 개수
                             ngram_range=(4, 4),
                             max_features = 20000
                            )
vectorizer
pipeline = Pipeline([
    ('vect', vectorizer),
])


%time train_data_features = pipeline.fit_transform(df_train['review_clean'])
%time test_data_features = pipeline.fit_transform(df_test['review_clean'])

In [ ]:
# train test 분리
y_train = df['author'].values

In [ ]:
x_train = df[:54879]                #train을 위한 독립 320000개     
x_test = df[54879:]              #test을 위한 독립 80000개

In [ ]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [ ]:
x_train

In [ ]:
x_test

In [ ]:
y_train=y_train[:54879]

In [ ]:
x_train = x_train.astype('float32') 
y_train = y_train.astype('float32')
x_test = x_test.astype('float32')   

In [ ]:
p_val = np.zeros((x_train.shape[0], 5))
p_tst = np.zeros((x_test.shape[0], 5))
#(cv.split(),1)
# -  StratifiedKFold 경우, trn,y 독립과 종속(동일 분포를 위해) 인자
for i, (i_trn, i_val) in enumerate(cv.split(x_train, y_train), 1):
    print(f'training model for CV #{i}')
    clf = lgb.LGBMClassifier(objective='multiclass',
                             n_estimators=3000,
                             num_leaves=256,
                             learning_rate=0.009,
                             subsample=.8,
                             subsample_freq=1,
                             num_iterations=2000,
                             colsample_bytree=.6,
                             random_state=42,
                             n_jobs=-1)
    
    clf.fit(x_train[i_trn], y_train[i_trn],eval_set=[(x_train[i_val], y_train[i_val])], eval_metric='multiclass',early_stopping_rounds=15)
    
    p_val[i_val, :] = clf.predict_proba(x_train[i_val])
    #확률로 예측값 출력
    p_tst += clf.predict_proba(x_test) / 5
    #평균 효과를 위해 먼저 폴드 개수만큼 나누어 초기화

In [ ]:
p_tst

In [ ]:
print(f'Log loss (CV) : {log_loss(pd.get_dummies(y_train),p_tst):8.4f}')

In [ ]:
print(f'Log loss (CV) : {log_loss(pd.get_dummies(y_train),p_val):8.4f}')